### PR TITLE
Simplify reification of predicate in bytecode and native compilers

### PR DESCRIPTION
**Pull request by maximedenes**

### CODE REVIEW

**Comment by maximedenes** [comment]
Any takers? @ppedrot ? @silene ?

**Comment by ppedrot** [comment]
I stumbled upon this code already and had a hard time understanding what the `dep` flag was standing for. I still don't quite understand, so I'm not sure I should take this PR.

**Comment by ppedrot** [comment]
Also, why do we rely on the shape of the type of the predicate? We know statically the number of arguments, don't we?

**Comment by maximedenes** [comment]
> I stumbled upon this code already and had a hard time understanding what the dep flag was standing for. I still don't quite understand, so I'm not sure I should take this PR.

I can explain : in a previous life, the return predicate could have two different shapes, depending on whether the match was dependent or not. When it was, the predicate was like today. When not, the quantification over the scrutinee was dropped.

**Comment by maximedenes** [comment]
> Also, why do we rely on the shape of the type of the predicate? We know statically the number of arguments, don't we?

The number yes, but we need to build the type of the domains, not so sure if it would be easy to do otherwise.

**Comment by ppedrot** [comment]
> I can explain : in a previous life, the return predicate could have two different shapes, depending on whether the match was dependent or not. When it was, the predicate was like today. When not, the quantification over the scrutinee was dropped (i.e. was not there).

Can you point to the commit that removed this non-uniformity?

**Comment by herbelin** [comment]
> Can you point to the commit that removed this non-uniformity?

This quite old but maybe was it done only by steps (in the absence of a way to enforce the uniformity in the ML syntax). For instance, this was done in `indrec.ml` in 2004 (d63e8fab799).

**Comment by ppedrot** [comment]
@maximedenes Can you rebase so that I can merge?

**Comment by ppedrot** [comment]
@maximedenes ping!

**Comment by maximedenes** [comment]
Rebased.

File: pretyping/vnorm.ml [hunk 1]
```diff
@@ -313,8 +310,8 @@ and nf_predicate env sigma ind mip params v pT =
       let params = if Int.equal n 0 then params else Array.map (lift n) params in
       let dom = mkApp(mkIndU ind,Array.append params rargs) in
       let body = nf_vtype (push_rel (LocalAssum (name,dom)) env) sigma vb in
-      true, mkLambda(name,dom,body)
-    | _ -> false, nf_val env sigma v crazy_type
+      mkLambda(name,dom,body)
+    | _ -> nf_val env sigma v crazy_type
```
Comment (ppedrot) [inline pretyping/vnorm.ml]: Thinking about it, shouldn't this pattern be an `assert false`? The return predicate is supposed to always be dependent now so there shouldn't be a shortage of lambdas.

**Comment by ppedrot** [comment]
Now there is a linting error.

**Review by mattam82 — APPROVED** [review]
I had no idea there was a time where the predicate depended on the dependency. Ok with the changes.

File: pretyping/vnorm.ml (hunk 1 above)
Comment (ppedrot) [reply pretyping/vnorm.ml]: Come on, you fixed the linting but didn't set the assertion failure?

Comment (maximedenes) [reply pretyping/vnorm.ml]: I can look at the code but not now. Fixing the whitespace didn't require opening the file.

**Comment by ppedrot** [comment]
I am waiting for the addition of the assertion failure mentioned above. If you really don't have time @maximedenes, I can merge it as-is nonetheless, I am just a bit concerned that we will forget about this if so. This is just a cleanup so there is no actual hurry.

**Pull request merged**